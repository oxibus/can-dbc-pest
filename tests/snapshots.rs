#![cfg(feature = "encoding")]

use std::fs;
use std::path::Path;

use can_dbc_pest::{decode_cp1252, DbcParser, Rule};
use insta::{assert_debug_snapshot, with_settings};
use pest::Parser;

/// Test parsing all DBC files
#[test]
fn test_cantools_dbc_files() {
    let test_dirs = [("tests/shared-test-files/dbc-cantools", "dbc-cantools")];
    for (path, file_name) in &test_dirs {
        let path = Path::new(path);
        let snapshot_path = String::from("snapshots-") + file_name;
        // snapshots should go to /tests/snapshots-* directory
        with_settings! {
            { omit_expression => true,
              snapshot_path => snapshot_path ,
              prepend_module_to_snapshot => false },
            {
                test_dbc_files(path);
            }
        }
    }
}

/// Test parsing all DBC files in the given directory.
fn test_dbc_files(dir: impl AsRef<Path>) {
    let dir = dir.as_ref();
    let dir_display = dir.display();
    let dir_content = fs::read_dir(dir).unwrap_or_else(|e| {
        panic!(
            "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
        )
    });

    eprintln!("Testing dbc files in directory: {dir_display}");
    for dbc_path in dir_content {
        let path = dbc_path.unwrap().path();
        if path
            .extension()
            .is_some_and(|ext| ext.eq_ignore_ascii_case("dbc"))
        {
            parse_one_file(path);
        }
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: impl AsRef<Path>) {
    let path = path.as_ref();
    eprintln!("Testing DBC file: {}", path.display());
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    if let Some(buffer) = decode_cp1252(&buffer) {
        match DbcParser::parse(Rule::file, &buffer) {
            Ok(pairs) => assert_debug_snapshot!(file_name, pairs),
            Err(e) => panic!("Failed to parse {file_name}.dbc: {e:#?}"),
        }
    } else {
        panic!("Failed to decode {file_name}.dbc as cp1252");
    }
}
