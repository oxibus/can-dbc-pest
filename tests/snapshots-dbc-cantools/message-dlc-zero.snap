---
source: tests/snapshots.rs
---
[
    Pair {
        rule: file,
        span: Span {
            str: "VERSION \"\"\n\nBS_:\n\nBU_: Node TestNode\n\n\nBO_ 1024 Message1: 0 TestNode\n\n\n\n\n",
            start: 0,
            end: 73,
        },
        inner: [
            Pair {
                rule: line,
                span: Span {
                    str: "VERSION \"\"",
                    start: 0,
                    end: 10,
                },
                inner: [
                    Pair {
                        rule: version,
                        span: Span {
                            str: "VERSION \"\"",
                            start: 0,
                            end: 10,
                        },
                        inner: [
                            Pair {
                                rule: string,
                                span: Span {
                                    str: "\"\"",
                                    start: 8,
                                    end: 10,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 10,
                    end: 11,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 11,
                    end: 12,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BS_:",
                    start: 12,
                    end: 16,
                },
                inner: [
                    Pair {
                        rule: bit_timing,
                        span: Span {
                            str: "BS_:",
                            start: 12,
                            end: 16,
                        },
                        inner: [],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 16,
                    end: 17,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 17,
                    end: 18,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BU_: Node TestNode",
                    start: 18,
                    end: 36,
                },
                inner: [
                    Pair {
                        rule: nodes,
                        span: Span {
                            str: "BU_: Node TestNode",
                            start: 18,
                            end: 36,
                        },
                        inner: [
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Node",
                                    start: 23,
                                    end: 27,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "TestNode",
                                    start: 28,
                                    end: 36,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 36,
                    end: 37,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 37,
                    end: 38,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 38,
                    end: 39,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "BO_ 1024 Message1: 0 TestNode",
                    start: 39,
                    end: 68,
                },
                inner: [
                    Pair {
                        rule: message,
                        span: Span {
                            str: "BO_ 1024 Message1: 0 TestNode",
                            start: 39,
                            end: 68,
                        },
                        inner: [
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "1024",
                                    start: 43,
                                    end: 47,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "Message1",
                                    start: 48,
                                    end: 56,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: integer,
                                span: Span {
                                    str: "0",
                                    start: 58,
                                    end: 59,
                                },
                                inner: [],
                            },
                            Pair {
                                rule: identifier,
                                span: Span {
                                    str: "TestNode",
                                    start: 60,
                                    end: 68,
                                },
                                inner: [],
                            },
                        ],
                    },
                ],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 68,
                    end: 69,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 69,
                    end: 70,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 70,
                    end: 71,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 71,
                    end: 72,
                },
                inner: [],
            },
            Pair {
                rule: line,
                span: Span {
                    str: "\n",
                    start: 72,
                    end: 73,
                },
                inner: [],
            },
            Pair {
                rule: EOI,
                span: Span {
                    str: "",
                    start: 73,
                    end: 73,
                },
                inner: [],
            },
        ],
    },
]
